File: 240407/조삼모사/three-at-dawn-and-four-at-dusk.cpp
#include <iostream>
#include <cstring>
#include <vector>
#include <set>
#include <climits>
#define fastio ios_base::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
using namespace std;
constexpr size_t MAX_N = 20;
int n, rst;
int arr[MAX_N][MAX_N];
vector<int> v;
set<int> s;

void init(){
    memset(arr, 0, sizeof(arr));
    rst = INT_MAX;
    v.push_back(0);
    s.insert(0);
}

void input(){
    cin >> n;
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < n; ++j) {
            cin >> arr[i][j];
        }
    }
}

int intensity(const vector<int> & data){
    int res = 0;
    for (int i = 0; i < n / 2; ++i) {
        for (int j = i + 1; j < n / 2; ++j) {
            res += arr[data[i]][data[j]];
            res += arr[data[j]][data[i]];
        }
    }

    return res;
}

void dfs(int idx, int depth){
    if(depth == n / 2){
        vector<int> tmp;
        for(int i = 0; i < n; i++){
            if(s.find(i) == s.end()) tmp.push_back(i);
        }

        rst = min(rst, abs(intensity(v) - intensity(tmp)));
    }

    for (int i = idx + 1; i < n; ++i) {
        s.insert(i);
        v.push_back(i);
        dfs(i, depth + 1);
        s.erase(i);
        v.pop_back();
    }
}

void run(){
    dfs(0, 1);
    cout << rst << endl;
}


int main() {
    fastio
    init();
    input();
    run();
    return 0;
}

File: 240407/조삼모사/three-at-dawn-and-four-at-dusk.py
# 강도 구하기
def intensity(c):
    res = 0
    for i in range(n//2):
        for j in range(i+1, n//2):
            res += p[c[i]][c[j]]
            res += p[c[j]][c[i]]

    return res

# 최솟값 구하기
def dfs(idx, depth):
    global ans

    if depth == n//2:
        tmp2 = [i for i in range(n) if i not in tmp]
        ans = min(ans, abs(intensity(tmp) - intensity(tmp2)))

    for i in range(idx+1, n):
        tmp.append(i)
        dfs(i, depth+1)
        tmp.pop()


if __name__ == '__main__':
    n = int(input())

    # 상성 리스트 저장
    p = []
    for _ in range(n):
        p.append(list(map(int, input().split())))

    tmp = [0]
    ans = float("INF")
    dfs(0, 1)
    print(ans)